tsl/src/compression/compression.c

```diff
@@ -2005,82 +2130,120 @@ decompress_batches_for_insert(ChunkInsertState *cis, Chunk *chunk, TupleTableSlo
-		 compressed_tuple = heap_getnext(heapScan, ForwardScanDirection))
+	if (OidIsValid(index_oid))
+	{
+		idxrel = index_open(index_oid, ExclusiveLock);
+		scankeys = fetch_index_scankeys(idxrel,
+										uncompressed_chunk->hypertable_relid,
+										slot,
+										&num_scankeys,
+										SEGMENTBY_KEYS);
+		index_scan = index_beginscan(in_rel, idxrel, GetLatestSnapshot(), num_scankeys, 0);
+		index_rescan(index_scan, scankeys, num_scankeys, NULL, 0);
+		index_slot = table_slot_create(in_rel, NULL);
+		/* fetch matching compressed tuple based on segmentby column values */
+		while (index_getnext_slot(index_scan, ForwardScanDirection, index_slot))
+		{
+			HeapTuple compressed_tuple = ExecFetchSlotHeapTuple(index_slot, false, NULL);
+			heap_deform_tuple(compressed_tuple,
+							  decompressor.in_desc,
+							  decompressor.compressed_datums,
+							  decompressor.compressed_is_nulls);
+			write_logical_replication_msg_decompression_start();
+			row_decompressor_decompress_row(&decompressor, NULL);
+			write_logical_replication_msg_decompression_end();
+
+			TM_FailureData tmfd;
+			TM_Result result pg_attribute_unused();
+			result = table_tuple_delete(in_rel,
+										&compressed_tuple->t_self,
+										decompressor.mycid,
+										GetTransactionSnapshot(),
+										InvalidSnapshot,
+										true,
+										&tmfd,
+										false);
+		}
+		ExecDropSingleTupleTableSlot(index_slot);
+		index_endscan(index_scan);
+		index_close(idxrel, ExclusiveLock);
+	}
+	else
 	{
-		Assert(HeapTupleIsValid(compressed_tuple));
-		bool valid = true;
-
-		/*
-		 * Since the heap scan API does not support SK_SEARCHNULL we have to check
-		 * for NULL values manually when those are part of the constraints.
-		 */
-		for (int attno = bms_next_member(null_columns, -1); attno >= 0;
-			 attno = bms_next_member(null_columns, attno))
+		scankeys = build_scankeys(uncompressed_chunk->fd.hypertable_id,
+								  uncompressed_chunk->hypertable_relid,
+								  decompressor,
+								  key_columns,
+								  &null_columns,
+								  slot,
+								  &num_scankeys);
+		bms_free(key_columns);
+		heapScan = table_beginscan(in_rel, GetLatestSnapshot(), num_scankeys, scankeys);
+
+		for (HeapTuple compressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
+			 compressed_tuple != NULL;
+			 compressed_tuple = heap_getnext(heapScan, ForwardScanDirection))
 		{
-			if (!heap_attisnull(compressed_tuple, attno, decompressor.in_desc))
+			Assert(HeapTupleIsValid(compressed_tuple));
+			bool valid = true;
+
+			/*
+			 * Since the heap scan API does not support SK_SEARCHNULL we have to check
+			 * for NULL values manually when those are part of the constraints.
+			 */
+			for (int attno = bms_next_member(null_columns, -1); attno >= 0;
+				 attno = bms_next_member(null_columns, attno))
 			{
-				valid = false;
-				break;
+				if (!heap_attisnull(compressed_tuple, attno, decompressor.in_desc))
+				{
+					valid = false;
+					break;
```

Comment:
Could you add a test for this line?

Reply:
tsl/test/sql/compression_conflicts.sql has test which will cover these lines.

Reply:
It seems Codecov still does not see any coverage here. Which SQL statement from the test should cover these lines?